sdk/python/lib/pulumi/runtime/resource_cycle_breaker.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2016-2021, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Set, Iterable
+
+if TYPE_CHECKING:
+    from .. import Resource
+
+
+_DEPENDENCIES_PROPERTY = '_direct_computed_dependencies'
```

Comment:
So this algorithm needs to store the state somewhere. Options to put the state in:

- some property of Resource objects (currently implemented)
- global table (I'm ever-so-slightly worried about leaks and re-entrancy in some auto API cases)
- table scoped to the Pulumi program run that is executing - this would perhaps be ideal, but where do we put such state, which object/class?

Reply:
@pgavlin has talked about adding a kv store as in the engine a first class feature in the pulumi RPC interface to service scenarios just like this. But I don't personally have anything against the approach you've taken to solve the customer issue right now.

Reply:
Yeah! Hm. This would indeed come handy one day. However as things stand with Py SDK I'd be a bit worried about relying on async KV store for core functionality.. For example I'd love to first get rid of threading and have a really foolproof async situation. in current design core functionality of Py SDK is to order the RegisterResource calls "right" before they hit the engine.

So far I can think of no downside piggybacking on the Resource prop - we're not ever pickling these objects right.. But if needed/wanted I can also explicitly declare this prop on the Resource class so it's more official this is being used.